### PR TITLE
update form rules.

### PR DESCRIPTION
**Pull request by Alice1319**
fixes eclipse-xpanse/xpanse #902
image

### CODE REVIEW

File: src/components/content/order/formElements/NumberInput.tsx [hunk 1]
```diff
@@ -5,23 +5,40 @@
...
     onChangeHandler: NumberInputEventHandler;
-}): JSX.Element {
+}): React.JSX.Element {
+    const ruleItems: Rule[] = [{ required: item.mandatory }, { type: 'number' }, { min: 0 }];
+    let minimum: number;
+    let maximum: number;
+
+    const setValidator = (_: unknown, value: number) => {
+        if (value <= minimum) {
+            return Promise.reject(new Error(`The new password not match minimum ${minimum}!`));
```
Comment (swaroopar) [inline src/components/content/order/formElements/NumberInput.tsx:24]: it is not always a password field. We need a better error message here.
`Value does not satisfy criteria minimum ${minimum}`

File: src/components/content/order/formElements/NumberInput.tsx [hunk 2]
```diff
@@ -5,23 +5,40 @@
...
+        }
+        if (value >= maximum) {
+            return Promise.reject(new Error(`The new password not match maximum ${maximum}!`));
```
Comment (swaroopar) [inline src/components/content/order/formElements/NumberInput.tsx:27]: `Value does not satisfy criteria maxinum ${maximum}`

File: src/components/content/order/formElements/NumberInput.tsx [hunk 3]
```diff
@@ -5,23 +5,40 @@
...
+        }
+        return Promise.reject(new Error('The new password that you entered do not match!'));
```
Comment (swaroopar) [inline src/components/content/order/formElements/NumberInput.tsx:29]: not a valid condition at all.

File: src/components/content/order/formElements/NumberInput.tsx [hunk 4]
```diff
@@ -5,23 +5,40 @@
...
+    };
+
+    for (const key in item.valueSchema) {
+        if (key === 'minimum' || key === 'maximum') {
```
Comment (swaroopar) [inline src/components/content/order/formElements/NumberInput.tsx:33]: can you please define a new enum type which has all supported json schema keys.. and then use it here. 
enum must have minimum, maximum, minimumLength, etc.. as allowed values.

File: src/components/content/order/formElements/SelectField.tsx [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import { cspMap } from './CspSelect';
+
+export function SelectField({
+    item,
+    onChangeHandler,
+}: {
+    item: DeployParam;
+    onChangeHandler: SelectOnChangeHandler;
+}): React.JSX.Element {
+    const activeCspList: CredentialVariables.csp[] = [
```
Comment (swaroopar) [inline src/components/content/order/formElements/SelectField.tsx:18]: please remove this definition of eum here again. 
Also the selectField must handle any enum values. Not just Csp

File: src/components/content/order/formElements/TextInput.tsx [hunk 6]
```diff
@@ -8,23 +8,46 @@ import { EyeInvisibleOutlined, EyeTwoTone, InfoCircleOutlined } from '@ant-desig
...
     item: DeployParam;
     onChangeHandler: TextInputEventHandler;
 }): React.JSX.Element {
+    const ruleItems: Rule[] = [{ required: item.mandatory }, { type: 'string' }];
+    let regExp: RegExp;
+    let minLength: number;
+    let maxLength: number;
+    const setValidator = (_: unknown, value: string) => {
+        if (value.length >= minLength && value.length <= maxLength && !regExp.test(value)) {
+            return Promise.reject(new Error(`The new password not match the pattern!${regExp}`));
```
Comment (swaroopar) [inline src/components/content/order/formElements/TextInput.tsx:25]: error text  is not correct. Better we execute each rule separately and return separate errors for minLength, maxLength and pattern.

**Review by swaroopar — CHANGES_REQUESTED** [review]

File: src/components/content/order/formElements/TextInput.tsx [hunk 7]
```diff
@@ -5,26 +5,42 @@
...
     onChangeHandler,
 }: {
     item: DeployParam;
     onChangeHandler: TextInputEventHandler;
 }): React.JSX.Element {
+    const ruleItems: Rule[] = [{ required: item.mandatory }, { type: 'string' }];
+    let regExp: RegExp;
+    const setValidator = (_: unknown, value: string) => {
+        if (!regExp.test(value)) {
+            return Promise.reject(new Error(`The new password not match the pattern!${regExp}`));
```
Comment (swaroopar) [inline src/components/content/order/formElements/TextInput.tsx:23]: also there should be space beween pattern! and the regExp

File: src/components/content/order/formElements/NumberInput.tsx [hunk 8]
```diff
@@ -3,25 +3,45 @@
...
     onChangeHandler: NumberInputEventHandler;
-}): JSX.Element {
+}): React.JSX.Element {
+    const ruleItems: Rule[] = [{ required: item.mandatory }, { type: 'number' }, { min: 0 }];
+    let minimum: number;
+    let maximum: number;
+
+    const setValidator = (_: unknown, value: number) => {
+        if (value <= minimum) {
+            return Promise.reject(new Error(`Value does not satisfy criteria minimum ${minimum}!`));
```
Comment (swaroopar) [inline src/components/content/order/formElements/NumberInput.tsx:24]: currently the error is shown multiple times.. also has some "undefined" error.

image

**Review by swaroopar — CHANGES_REQUESTED** [review]

File: src/components/content/order/formElements/CommonTypes.ts [hunk 9]
```diff
@@ -29,6 +36,7 @@ export interface DeployParam {
     value: string;
     mandatory: boolean;
     sensitiveScope: DeployVariable.sensitiveScope;
+    valueSchema: Record<string, Record<string, unknown>> | undefined;
```
[the remark this answers is not in the document]
Comment (Alice1319) [reply src/components/content/order/formElements/CommonTypes.ts:38]: of course, we can do it

**Pull request closed without merging**